parser/prism/Translator.cc

```diff
@@ -134,6 +134,97 @@ ast::ExpressionPtr Translator::desugarDString(core::LocOffsets loc, pm_node_list
                     static_cast<uint16_t>(interpArgs.size()), move(interpArgs));
 }
 
+// Desugar multiple left hand side assignments into a sequence of assignments
+//
+// Considering this example:
+// ```rb
+// arr = [1, 2, 3]
+// a, *b = arr
+// ```
+//
+// We desugar the assignment `a, *b = arr` into:
+// ```rb
+// tmp = ::<Magic>.expandSplat(arr, 1, 0)
+// a = tmp[0]
+// b = tmp.to_ary
+// ```
+//
+// While calling `to_ary` doesn't return the correct value if we were to execute this code,
+// it returns the correct type from a static point of view.
+ast::ExpressionPtr Translator::desugarMlhs(core::LocOffsets loc, parser::Mlhs *lhs, ast::ExpressionPtr rhs) {
+    ast::InsSeq::STATS_store stats;
+
+    core::NameRef tempRhs = nextUniqueDesugarName(core::Names::assignTemp());
+    core::NameRef tempExpanded = nextUniqueDesugarName(core::Names::assignTemp());
+
+    int i = 0;
+    int before = 0, after = 0;
+    bool didSplat = false;
+    auto zloc = loc.copyWithZeroLength();
+
+    for (auto &c : lhs->exprs) {
+        if (auto *splat = parser::NodeWithExpr::cast_node<parser::SplatLhs>(c.get())) {
```

Comment:
Oh boy, I have no idea how we'd do this if the wq tree didn't exist. I'll make an issue to revisit.